src/main/java/com/google/finapp/SpannerDaoImpl.java

```diff
@@ -125,26 +128,35 @@ public void moveAccountBalance(ByteArray fromAccountId, ByteArray toAccountId, B
                     readAccountBalances(fromAccountId, toAccountId, transaction);
 
                 BigDecimal newSourceAmount = accountBalances.get(fromAccountId).subtract(amount);
+                BigDecimal newDestAmount = accountBalances.get(toAccountId).add(amount);
 
                 if (newSourceAmount.signum() == -1) {
                   throw new IllegalArgumentException(
                       String.format(
-                          "Account balance cannot be negative. original account balance: %s, amount transferred: %s",
+                          "Cannot transfer amount greater than original balance. fromAccount balance: %s, amount: %s",
                           accountBalances.get(fromAccountId).toString(), amount.toString()));
                 }
 
                 transaction.buffer(
                     ImmutableList.of(
                         buildUpdateAccountMutation(fromAccountId, newSourceAmount),
-                        buildUpdateAccountMutation(
-                            toAccountId, accountBalances.get(toAccountId).add(amount)),
+                        buildUpdateAccountMutation(toAccountId, newDestAmount),
                         buildInsertTransactionHistoryMutation(
                             fromAccountId, amount, /* isCredit= */ true),
                         buildInsertTransactionHistoryMutation(
                             toAccountId, amount, /* isCredit= */ false)));
+
+                accountBalancesBuilder.put(fromAccountId, newSourceAmount);
+                accountBalancesBuilder.put(toAccountId, newDestAmount);
                 return null;
               });
+      return accountBalancesBuilder.build();
     } catch (SpannerException e) {
+      // filter for IllegalArgumentExceptions thrown in lambda function above
+      Throwable cause = e.getCause();
+      if (cause instanceof IllegalArgumentException) {
+        throw new IllegalArgumentException(cause.getMessage());
+      }
```

Comment:
Seems as though exceptions thrown in the lambda function end up being wrapped by `SpannerException`. The point of the filter is to give more specific/useful feedback for the user. However, this filtering can be moved to `FinAppService`, which is what calls this function. I chose to filter here, though, so you can see directly where the `IllegalArgumentException` is coming from and why we filter through `SpannerException`s

Reply:
Actually, this needs to be in this location in order for the two spanner dao implementations to have identical expected outcomes for invalid inputs. Moving the filter elsewhere will break the tests as I have them setup--which expects `IllegalArgumentException` for illegal arguments. However, this isn't in the signature, so it's debatable if I should even be testing this.

Reply:
#34